kotlinx-coroutines-core/common/src/flow/internal/DistinctFlow.kt

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2016-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow.internal
+
+import kotlinx.coroutines.flow.*
+
+/**
+ * An internal interface that marks the flow impls with distinct values, so that
+ * application of [distinctUntilChanged] can be optimized away. A class implementing
+ * this interface can be conditionally distinct via [isDefaultEquivalence] without
+ * having to have multiple copies of the same code.
+ */
+internal interface DistinctFlow<T> : Flow<T> {
```

Comment:
The key motivation was based around the logic of `ShartingStarted.command`. As `SharingStarted` documentation states "Back-to-back emissions of the same command have no effect.", so to implement it, implementation of `shareIn` operator applies `distinctUntilChanged`:
https://github.com/Kotlin/kotlinx.coroutines/blob/shared-flow/kotlinx-coroutines-core/common/src/flow/operators/Share.kt#L203-L205
However, actual default implementations of `command` function return flows that are already distinct by definition, and applying yet another operator to them is not simply a waste, it adds an additional allocated object for every active instance of `shareIn` operator. Adding a simple interface seems like a small price to fix it, since we expect `shareIn` to be an actively used operator. (more code, yet less runtime overhead).

Reply:
Now, I've implemented another (better) optimization for SharingStarted.Eagerly/Lazily, so this DistinctFlow optimization is no longer needed, removed it.